src/backend/api/handlers.py

```diff
@@ -0,0 +1,191 @@
+# coding=utf-8
+import math
+import os
+from typing import AnyStr
+
+import dateparser
+
+from api.exceptions import BadRequestFormatExceptioin
+from api.mirrors_update import (
+    get_config,
+    get_verified_mirrors,
+    REQUIRED_MIRROR_PROTOCOLS,
+)
+from db.db_engine import GeoIPEngine
+from db.models import Url, Mirror
+from db.utils import session_scope
+from sqlalchemy.sql.expression import false
+
+from common.sentry import (
+    get_logger,
+)
+
+
+MAX_LENGTH_OF_MIRRORS_LIST = 5
+
+logger = get_logger(__name__)
+
+
+def _get_distance_between(
+        lat1: float,
+        lon1: float,
+        lat2: float,
+        lon2: float,
+) -> float:
+    return math.sqrt(
+        (lat1 - lat2)**2 + (lon1 - lon2)**2
+    )
+
+
+def _get_nearest_mirrors(ip_address):
+    db = GeoIPEngine.get_instance()
+    match = db.lookup(ip_address)
+    with session_scope() as session:
+        if match is None:
+            all_mirrors = session.query(Mirror).filter(
+                Mirror.is_expired == false(),
+                ).all()
+            all_mirrors = [mirror.to_dict() for mirror in all_mirrors]
+            return all_mirrors
+        match_dict = match.get_info_dict()
+        match_country = match_dict['country']['names']['en']
+        match_continent = match_dict['continent']['names']['en']
+        suitable_mirrors = session.query(Mirror).filter(
+            Mirror.continent == match_continent,
+            Mirror.country == match_country,
+            Mirror.is_expired == false(),
+        ).limit(MAX_LENGTH_OF_MIRRORS_LIST).all()
+        suitable_mirrors = [mirror.to_dict() for mirror in suitable_mirrors]
+        if len(suitable_mirrors) == MAX_LENGTH_OF_MIRRORS_LIST:
+            return suitable_mirrors
+        rest_mirrors = session.query(Mirror).filter(
+            Mirror.continent == match_continent,
+            Mirror.is_expired == false(),
+            Mirror.name.notin_(
+                [mirror.name for mirror in suitable_mirrors]
+            ),
+        ).all()
+        rest_mirrors = [mirror.to_dict() for mirror in rest_mirrors]
+        if not rest_mirrors:
+            rest_mirrors = session.query(Mirror).filter(
+                Mirror.name.notin_(
+                    [mirror.name for mirror in suitable_mirrors]
+                ),
+                Mirror.is_expired == false(),
+            ).all()
+            rest_mirrors = [mirror.to_dict() for mirror in rest_mirrors]
+        distance_dict = {
+            _get_distance_between(
+                lat1=match.location[0],
```

Comment:
Btw you gave me an idea 

Django ORM is able to return calculated field, so can SQLAlchemy I guess. You can move this ordering and sorting to postgresql (or even check postgis, it has better indexes and methods for lat/lon mathmatics). That will be much faster with indexes than python code.

May be postponed to the future updates.

Reply:
Done. Now it's used hybrid method.